### PR TITLE
Define compiler variables in CMakePresets.json

### LINKED ISSUE #13136: [feature] CMakePresets support for msvc toolset version
## Suggestion

Would it be possible for the Conan generated CMakePresets.json to include `msvc.toolset` version details (i.e. `v143`, `v142`, etc.) when populating the `toolset.value` field?

This would allow IDEs to know which `vcvarsall.bat` environment to use when building the project.

## Expected Behavior

For example, when cross-compiling with VS2022 x64 to x86, I'd expect to see the conan generated CMakePresets.json look something like:
```diff
"generator": "Ninja",
"cacheVariables": {
    "CMAKE_POLICY_DEFAULT_CMP0091": "NEW",
    "CMAKE_BUILD_TYPE": "RelWithDebInfo"
},
"architecture": {
    "value": "x86",
    "strategy": "external"
},
+"toolset": {
+    "value": "v143,host=x64",
+    "strategy": "external"
+},
"toolchainFile": "C:\\projects\\build\\foo\\RelWithDebInfo\\generators\\conan_toolchain.cmake",
"binaryDir": "C:\\projects\\build\\foo\\RelWithDebInfo
```            

_I can use `tools.cmake.cmaketoolchain:toolset_arch=x64` to add the "host=x64" portion, but I don't have anyway to get `v143` added._

VSCode recognizes this `toolset` / `architecture` combination and correctly chooses VS2022 and calls `vcvarsall.bat amd64_x86` prior to calling CMake.

## Current Behavior

Conan *doesn't* include `toolset` details in the generated CMakePresets.json file, so in order to get IDEs like VSCode to find the right environment, we need to adjust CMakePresets.json (as described above) or worse adjust the CMakeUserPresets.json file as follows:

```diff
{
    "version": 4,
    "vendor": {
        "conan": {}
    },
    "include": [
        "C:\\path\\to\\conan\\CMakePresets.json"
-    ]
+    ],
+    "configurePresets": [
+      {
+        "name": "RelWithDebInfo",
+        "displayName": "RelWithDebInfo",
+        "inherits": "relwithdebinfo",
+        "toolset": {
+          "value": "v143,host=x64",
+          "strategy": "external"
+        }
+    ],
+    "buildPresets": [
+        {
+            "name": "RelWithDebInfo",
+            "displayName": "RelWithDebInfo",
+            "configurePreset": "RelWithDebInfo",
+            "configuration": "RelWithDebInfo"
+        }
+    ]
}
```

Conan already knows *which* vcvarsall to call, and and ensures the correct one gets called in the generated `conanvcvars.bat`.

Conan also appears to know about `msvc.toolset` version information:
https://github.com/conan-io/conan/blob/e33b55a486fd208223524dda49cce02dbe70c214/conans/client/conf/__init__.py#L117
_At least for the legacy toolsets_

And seems to know about the mapping:
https://github.com/conan-io/conan/blob/e0a8ee058bc8dc2d9811b8aeb6999f69aeb78d85/conan/tools/microsoft/visual.py#L53-L60

It seems like Conan could should be able to use the `msvc.toolset` information from the build profile and populate the version field correctly.

## Context
Follow up to https://github.com/conan-io/conan/issues/11623

Some IDEs are capable of targeting multiple toolsets (e.g. VS2015, VS2019, VS2022), and `toolset.value` is the way for the IDE to know which `vcvarsall.bat` to load and use internally while compiling.

> It's the same step that Visual Studio takes for you when the IDE invokes CMake. **Visual Studio parses the active Configure Preset for the host and target architecture specified by toolset and architecture. Visual Studio then sources the specified environment from vcvarsall.bat.** When you build from the Windows command line with Ninja, you'll need to take this step yourself.
_Excerpt from MSDN [Sourcing the environment when building with command-line generators on Windows](https://learn.microsoft.com/en-us/cpp/build/cmake-presets-vs?view=msvc-170#sourcing-the-environment-when-building-with-command-line-generators-on-windows)_

IDE | Support 
----|----------
VSCode | [Already Supported](https://github.com/microsoft/vscode-cmake-tools/search?q=varsForVSInstallation) <br> https://github.com/microsoft/vscode-cmake-tools/pull/2524
QtCreator | [In Progress](https://codereview.qt-project.org/c/qt-creator/qt-creator/+/457588/8/src/plugins/cmakeprojectmanager/cmakeprojectimporter.cpp)<br> [QTCREATOR-BUG 28693](https://bugreports.qt.io/browse/QTCREATORBUG-28693)
CLion | [Feature Request](https://youtrack.jetbrains.com/issue/CPP-31353/Support-external-strategy-of-toolset-and-architecture-fields-in-CMakePresets.json-for-MSVC-compilers.)

Thanks for your support!

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### PR DESCRIPTION
**Pull request by petwu**
Changelog: Feature: Define `CMAKE_<LANG>_COMPILER` variables in CMakePresets.json.
Docs: Omit
Fixes #13136

The compiler variables are set to the same values as in the `cmake_toolchain.cmake`, and only when defining `tools.build:compiler_executables`.

I've tested it with both VS an VS Code, which successfully initialize the correct vcvars environment and select the correct compiler as specified by the toolset version, host and target architecture. From a non-IDE perspective, this does influence the CMake configuration in any way, as the compiler variables would be set in the toolchain to the same value anyway.

<details><summary>profile</summary>

```ini
[settings]
os=Windows
arch=x86
compiler=msvc
compiler.version=193
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.cppstd=17

[conf]
tools.cmake.cmaketoolchain:toolset_arch=x64
tools.build:compiler_executables={"c": "cl.exe", "cpp": "cl.exe"}
```

</details>

<details><summary>conanfile.txt</summary>

```ini
[generators]
CMakeToolchain

[layout]
cmake_layout
```

</details>

<details><summary>CMakePresets.json after <code>conan install -pr profile -s build_type=Debug .</code></summary>

```json
{
    "version": 3,
    "vendor": {
        "conan": {}
    },
    "cmakeMinimumRequired": {
        "major": 3,
        "minor": 15,
        "patch": 0
    },
    "configurePresets": [
        {
            "name": "conan-debug",
            "displayName": "'conan-debug' config",
            "description": "'conan-debug' configure using 'Ninja' generator",
            "generator": "Ninja",
            "cacheVariables": {
                "CMAKE_POLICY_DEFAULT_CMP0091": "NEW",
                "CMAKE_BUILD_TYPE": "Debug",
                "CMAKE_C_COMPILER": "cl.exe",
                "CMAKE_CXX_COMPILER": "cl.exe"
            },
            "toolset": {
                "value": "v143,host=x64",
                "strategy": "external"
            },
            "architecture": {
                "value": "x86",
                "strategy": "external"
            },
            "toolchainFile": "generators\\conan_toolchain.cmake",
            "binaryDir": "C:\\TMP\\conan-feat-presets-with-compiler\\build\\Debug"
        }
    ],
    "buildPresets": [
        {
            "name": "conan-debug",
            "configurePreset": "conan-debug",
            "jobs": 24
        }
    ],
    "testPresets": [
        {
            "name": "conan-debug",
            "configurePreset": "conan-debug"
        }
    ]
}
```

</details> 

- [x] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

**Review by memsharded — APPROVED** [review]
I think this can be moved forward for next release, thanks very much for  your contribution!

Asking @jcar87 for an extra review, thanks!

File: conan/tools/cmake/presets.py [hunk 1]
```diff
@@ -158,6 +158,13 @@ def _configure_preset(conanfile, generator, cache_variables, toolchain_file, mul
                     "strategy": "external"
                 }
 
+        # Set the compiler like in the toolchain. Some IDEs like VS or VSCode require the compiler
+        # being set to cl.exe in order to activate the environment using vcvarsall.bat according to
```
Comment (jcar87) [inline conan/tools/cmake/presets.py:162]: interesting! I've not found this documented for VS Code, but have verified that it indeed works that way. Thanks so much for figuring this out! :D

File: test/integration/toolchains/cmake/test_cmaketoolchain.py [hunk 2]
```diff
@@ -1046,6 +1046,30 @@ def test_set_cmake_lang_compilers_and_launchers():
     assert 'set(CMAKE_RC_COMPILER "C:/local/rc.exe")' in toolchain
 
 
+def test_cmake_presets_compiler():
+    profile = textwrap.dedent(r"""
+    [settings]
+    os=Windows
+    arch=x86_64
+    compiler=clang
```
Comment (jcar87) [inline test/integration/toolchains/cmake/test_cmaketoolchain.py]: I can see the output is as expected, but wouldn't he compiler executables be `clang` or `clang-cl.exe` instead?

**Review by jcar87 — APPROVED** [review]
LGTM!

@memsharded, I would consider passing `cl.exe` to both C and CXX compilers if `compiler=msvc`, the cmake generator is Ninja, _even when_ `tools.build.compiler_executables` is undefined.

File: test/integration/toolchains/cmake/test_cmaketoolchain.py (hunk 2 above)
Comment (memsharded) [reply test/integration/toolchains/cmake/test_cmaketoolchain.py]: I think this is being directly overriden by ``tools.build:compiler_executables`` and most likely the intention of the test was to have ``compiler=msvc``, can you please confirm @petwu?

I'll check if it makes sense or is possible to introduce some check for this.

Comment (memsharded) [reply test/integration/toolchains/cmake/test_cmaketoolchain.py]: I suspect this is derived from copy&paste the test above called ``test_set_cmake_lang_compilers_and_launchers``

Comment (memsharded) [reply test/integration/toolchains/cmake/test_cmaketoolchain.py]: Exactly, I have pushed some changes to the test to define ``compiler=msvc`` instead.
Indeed, the ``compiler_executables`` is being written no matter what the ``compiler`` setting is, it is the responsibility of the user to define the profiles consistently. I have checked if it is possible to introduce some check that made sense for potential misalignments like this one (defining ``compiler_executable`` with ``cl.exe`` for other compiler not ``msvc``, but it looks a bit too dirty, so leaving it out of this PR, might reconsider later.

File: conan/tools/cmake/presets.py (hunk 1 above)
Comment (petwu) [reply conan/tools/cmake/presets.py:162]: I thought it was documented somewhere, but I couldn't find it either and can't remember how I learned about it in the first place. But I have used this trick successfully for some time now and seeing [this comment](https://github.com/microsoft/vscode-cmake-tools/issues/1929#issuecomment-858800902) this has been a thing for at least 3 years.

File: test/integration/toolchains/cmake/test_cmaketoolchain.py (hunk 2 above)
Comment (petwu) [reply test/integration/toolchains/cmake/test_cmaketoolchain.py]: You are right, I just copied the test above. Setting `compiler=clang` and then saying your executable is called `cl.exe` of course doesn't make sense in a real-world scenario. The test should just verify that the `compiler_executable` setting ended up in the generated preset.

**Pull request merged**